### PR TITLE
[Kernels] Add support for hybrid DeepEP

### PR DESCRIPTION
**Pull request by bnellnm**
## Purpose

Add support for the hybrid DeepEP buffer to vllm.

See https://github.com/deepseek-ai/DeepEP/pull/420 and https://github.com/deepseek-ai/DeepEP/tree/hybrid-ep

Currently using nvshmem 3.4.5 and hash `ef73fd9` + https://github.com/deepseek-ai/DeepEP/pull/470 on the [hybrid-ep](https://github.com/deepseek-ai/DeepEP/tree/hybrid-ep) branch

I feel this is still a bit hacky.  Having to dispatch the `topk_ids` and `topk_weights` is not ideal.  And maybe the padding should be done at a different level.

Also note that the hybrid kernels do not work with cudagraphs as currently implemented due to uses of torch `all_gather` and `item()`.

## Test Plan

Add test to `test_modular_kernel_combinations.py`
lm-evals + benchmarks with DeepSeek model(s)

## Test Result

cc @varun-sundar-rabindranath , @wenscarl 

---
<details>
<summary> Essential Elements of an Effective PR Description Checklist </summary>

- [ ] The purpose of the PR, such as "Fix some issue (link existing issues this PR will resolve)".
- [ ] The test plan, such as providing test command.
- [ ] The test results, such as pasting the results comparison before and after, or e2e results
- [ ] (Optional) The necessary documentation update, such as updating `supported_models.md` and `examples` for a new model.
- [ ] (Optional) Release notes update. If your change is user facing, please update the release notes draft in the [Google Doc](https://docs.google.com/document/d/1YyVqrgX4gHTtrstbq8oWUImOyPCKSGnJ7xtTpmXzlRs/edit?tab=t.0).
</details>

### CODE REVIEW

File: vllm/model_executor/layers/fused_moe/config.py [hunk 1]
```diff
@@ -719,6 +719,10 @@ def use_deepep_ht_kernels(self):
     def use_deepep_ll_kernels(self):
         return self.use_all2all_kernels and self.all2all_backend == "deepep_low_latency"
 
+    @property
+    def use_deepep_hybrid_kernels(self):
+        return self.use_all2all_kernels and envs.VLLM_ALL2ALL_BACKEND == "deepep_hybrid"
```
Comment (tlrmchlsmth) [inline vllm/model_executor/layers/fused_moe/config.py]: The environment variable is deprecated
```suggestion
    @property
    def use_deepep_hybrid_kernels(self):
        return self.use_all2all_kernels and self.all2all_backend == "deepep_hybrid"
```

File: vllm/distributed/device_communicators/all2all.py [hunk 2]
```diff
@@ -320,6 +320,40 @@ def set_num_sms(self, num_sms: int):
...
+    def __init__(self, cpu_group):
+        assert has_hybrid_deep_ep(), "DeepEP Hybrid kernels not found."
+        super().__init__(cpu_group)
+
+    def _make_all2all_kwargs(self, **kwargs) -> dict[Any, Any]:
+        extra_kwargs = dict(
+            group=self.cpu_group,
+            # num_sms_dispatch_api=32,
+            # num_sms_combine_api=32,
+            # num_sms_preprocessing_api=128,
```
Comment (tlrmchlsmth) [inline vllm/distributed/device_communicators/all2all.py]: is this debug cruft? Remove if not needed, or add a comment

File: tests/kernels/moe/modular_kernel_tools/common.py [hunk 3]
```diff
@@ -577,7 +580,8 @@ def next_power_of_2(x):
         num_local_experts=config.num_local_experts,
         moe_parallel_config=moe_parallel_config,
         in_dtype=config.dtype,
-        max_num_tokens=next_power_of_2(config.M),
+        # 128 needed for hybrid DeepEP. TODO(bnell): make this smarter
+        max_num_tokens=max(128, next_power_of_2(config.M)),
```
Comment (tlrmchlsmth) [inline tests/kernels/moe/modular_kernel_tools/common.py]: Could you resolve the TODO before merging? Where does the 128 come from?

File: vllm/v1/worker/gpu_model_runner.py [hunk 4]
```diff
@@ -1316,7 +1316,10 @@ def _prepare_inputs(
         # running prefills. This lets us set enforce_eager on the prefiller in
         # a P/D setup and still use CUDA graphs (enabled by this padding) on the
         # decoder.
-        allow_dp_padding = self.compilation_config.cudagraph_mode != CUDAGraphMode.NONE
+        allow_dp_padding = (
+            self.compilation_config.cudagraph_mode != CUDAGraphMode.NONE
+            or self.parallel_config.all2all_backend == "deepep_hybrid"
+        )
```
Comment (tlrmchlsmth) [inline vllm/v1/worker/gpu_model_runner.py]: Do the hybrid DeepEP kernels support CUDA Graphs? I'm seeing conflicting things in the PR

File: vllm/model_executor/layers/fused_moe/deepep_hybrid_prepare_finalize.py [hunk 5]
```diff
@@ -0,0 +1,280 @@
...
+        )
+
+    def create_new_topk_data(
+        self,
+        topk_ids: torch.Tensor,
+        topk_weights: torch.Tensor,
+    ) -> tuple[torch.Tensor, torch.Tensor]:
+        all_topk_ids, all_topk_weights = get_dp_group().all_gatherv(
+            [topk_ids, topk_weights], dim=0
+        )
```
Comment (tlrmchlsmth) [inline vllm/model_executor/layers/fused_moe/deepep_hybrid_prepare_finalize.py:116]: We're doing a dense `all_gatherv` after calling `dispatch`? This is not going to scale well. Could you explain why we need to do this?

File: vllm/model_executor/layers/fused_moe/deepep_hybrid_prepare_finalize.py [hunk 6]
```diff
@@ -0,0 +1,280 @@
...
+    ):
+        super().__init__()
+        self.buffer = buffer
+        self.num_dispatchers_ = num_dispatchers
+        self.rank_expert_offset = rank_expert_offset
+        self.handle = None
+        self.expert_probs = None
+        self.num_local_experts = num_local_experts
+        # Set to True to try cudagraph support.
+        self.do_routing = False
```
Comment (tlrmchlsmth) [inline vllm/model_executor/layers/fused_moe/deepep_hybrid_prepare_finalize.py]: Could you explain what this flag does and how it relates to CUDA Graph support?

File: tests/kernels/moe/modular_kernel_tools/mk_objects.py [hunk 7]
```diff
@@ -89,6 +94,12 @@ class ExpertInfo:
 common_float_and_int_types = common_float_types + [torch.int8]
 nvfp4_types = ["nvfp4"]
 fp8_types = [torch.float8_e4m3fn]
+fp8_bf16_types = [torch.float8_e4m3fn, torch.bfloat16]
+
+# Use filters for testing specific classes.
+# 'None' allows all possible object types.
+PREPARE_FINALIZE_FILTER = None
+EXPERTS_FILTER = None
```
Comment (tlrmchlsmth) [inline tests/kernels/moe/modular_kernel_tools/mk_objects.py:101]: how do these filters work?

File: vllm/v1/worker/gpu_model_runner.py (hunk 4 above)
Comment (bnellnm) [reply vllm/v1/worker/gpu_model_runner.py]: They don't.

File: tests/kernels/moe/modular_kernel_tools/common.py (hunk 3 above)
Comment (bnellnm) [reply tests/kernels/moe/modular_kernel_tools/common.py]: The 128 is a size/alignment limitation of the hybrid kernels.  I can put it in a constant or something.

File: vllm/model_executor/layers/fused_moe/deepep_hybrid_prepare_finalize.py (hunk 5 above)
Comment (bnellnm) [reply vllm/model_executor/layers/fused_moe/deepep_hybrid_prepare_finalize.py:116]: Yeah, this is not ideal.  It was the only way I could get it to work though.  The deep gemm (and other kernels) require the dispatched topk ids and/or weights and I couldn't figure out how to reconstruct these from the various maps that the hybrid kernels returned.

File: vllm/model_executor/layers/fused_moe/deepep_hybrid_prepare_finalize.py (hunk 6 above)
Comment (bnellnm) [reply vllm/model_executor/layers/fused_moe/deepep_hybrid_prepare_finalize.py]: The hybrid kernels have the option of taking in a pre-computed routing map instead of computing it internally.  The internally computed map uses a torch all gather which breaks cudagraphs.  When `do_routing` is `True`, the routing map is computed by vLLM using its cudagraph-able communication primitives.  Unfortunately, there's other stuff in the hybrid kernels that break things.  I left this flag in for future work on cudagraph support.

I'll add some comments.

File: tests/kernels/moe/modular_kernel_tools/mk_objects.py (hunk 7 above)
Comment (bnellnm) [reply tests/kernels/moe/modular_kernel_tools/mk_objects.py:101]: I'll add a comment.  They are lists of string class names that are allowed to be registered.

File: vllm/model_executor/layers/fused_moe/config.py (hunk 1 above)
Comment (varun-sundar-rabindranath) [reply vllm/model_executor/layers/fused_moe/config.py]: I think it is better to have it and then remove all `VLLM_ALL2ALL_BACKEND` in one go when we remove it ?

File: vllm/model_executor/layers/fused_moe/deep_gemm_moe.py [hunk 8]
```diff
@@ -53,6 +54,9 @@ def _valid_deep_gemm(
         logger.debug_once("DeepGemm disabled: deep_gemm not available.")
         return False
 
+    if is_deep_gemm_e8m0_used():
+        return True
```
Comment (varun-sundar-rabindranath) [inline vllm/model_executor/layers/fused_moe/deep_gemm_moe.py]: should this be true only if `deepep_hybrid` is used ?

File: vllm/model_executor/layers/fused_moe/deepep_hybrid_prepare_finalize.py (hunk 5 above)
Comment (varun-sundar-rabindranath) [reply vllm/model_executor/layers/fused_moe/deepep_hybrid_prepare_finalize.py:116]: Is it possible to use expert_probs for inferring topk ids and topk weights ?

Not very sure about this, but it looks like expert_probs is a tensor of size [num_tokens, num_local_experts] and all non-zero indices are topk_ids and non-zero values are topk_weights.

File: vllm/model_executor/layers/fused_moe/fused_moe.py [hunk 9]
```diff
@@ -1597,7 +1596,7 @@ def fused_experts(
     if (
         allow_deep_gemm
         and quant_config.use_fp8_w8a8
-        and (is_deep_gemm_e8m0_used() or _valid_deep_gemm(hidden_states, w1, w2))
```
Comment (varun-sundar-rabindranath) [inline vllm/model_executor/layers/fused_moe/fused_moe.py:1645]: I see. This is just moved into the `_valid_deep_gemm` .

File: vllm/model_executor/layers/fused_moe/layer.py [hunk 10]
```diff
@@ -1789,6 +1819,22 @@ def process_chunk(chunk_start, chunk_end, skip_result_store=False):
...
+                    (0, 0, 0, pad),
+                    mode="constant",
+                    value=0.0,
+                )
+                staged_router_logits = F.pad(
+                    staged_router_logits,
+                    (0, 0, 0, pad),
+                    mode="constant",
+                    value=0.0,
+                )
```
Comment (varun-sundar-rabindranath) [inline vllm/model_executor/layers/fused_moe/layer.py]: I believe we just ignored these "invalid/garbage" inputs and outputs. Does hybrid_deepep fail if we dont do this ?

File: vllm/v1/worker/gpu_model_runner.py (hunk 4 above)
Comment (bnellnm) [reply vllm/v1/worker/gpu_model_runner.py]: They don't but need dp padding regardless.

File: vllm/model_executor/layers/fused_moe/deepep_hybrid_prepare_finalize.py (hunk 5 above)
Comment (bnellnm) [reply vllm/model_executor/layers/fused_moe/deepep_hybrid_prepare_finalize.py:116]: I've tried using expert_probs but I could not find a way to reconstruct the ordering so the results are gibberish.

File: vllm/model_executor/layers/fused_moe/layer.py (hunk 10 above)
Comment (bnellnm) [reply vllm/model_executor/layers/fused_moe/layer.py]: It needs specific sizes to work, otherwise it'll assert.

File: vllm/model_executor/layers/fused_moe/fused_moe.py (hunk 9 above)
Comment (bnellnm) [reply vllm/model_executor/layers/fused_moe/fused_moe.py:1645]: I'm going to revert these changes since they are unrelated.
